### DISCUSSION TITLE
where is webui.js?

### DISCUSSION THREAD

**Discussion by laoshaw** [webui-dev/webui#320] [Q&A]
all main.c refers to webui.js and it looks like it's built from bridge/webui_bridge.ts, however I can run the examples without building the typescript file at all, how did the various "main" examples get its webui.js then?

**Comment by fibodevy** [comment]
It's embedded in the webui lib as array of bytes. js2c.py saves it to webui_bridge.h

**Reply by ttytm** [reply]
Right it's a "virtual" file.

**Reply by melMass** [reply]
How do you use it in vitejs for instance?

**Reply by AlbertShown** [reply]
it should be at `http://localhost:` + `YOUR_WINDOW_PORT_NUMBER` + `/webui.js`.

**Reply by ttytm** [reply]
@melMass 

Sharing an example when using Vite with SvelteKit, the script line with the virtual webui.js file can be added to `src/app.html`:

```diff
<! -- app.html -->
...
	<meta name="viewport" content="width=device-width, initial-scale=1" />
+	<script src="webui.js"></script>
	%sveltekit.head%
...
```

For more control, it can also reside in `÷layout.svelte`:
```svelte
<! -- +layout.svelte -->
...
<svelte:head>
	<script src="webui.js"></script>
</svelte:head>
```

For example, this would allow to make it conditional. One case where this can be useful is if parts of the frontend are developed in a regular browser / in an environment without webui. Like:
```svelte
<! -- +layout.svelte -->
...
<svelte:head>
	{#if process.env.NODE_ENV === 'production'}
		<script src="webui.js"></script>
	{/if}
</svelte:head>
```

For other frameworks just adapt accordingly where you can throw in that script line.

**Reply by melMass** [reply]
Thanks for the detailed reply! Here is what I ended up doing for [rsbuild](https://github.com/web-infra-dev/rsbuild):

**src/template.html**
```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  </head>
  <body>
    <script src="/webui.js"></script>
  </body>
</html>
```

**rsbuild.config.ts**
```ts
import { defineConfig } from "@rsbuild/core";
import { pluginSvelte } from "@rsbuild/plugin-svelte";

export default defineConfig({
  plugins: [pluginSvelte()],
  html: {
    inject: "body",
    template: "./src/template.html"
    },
  },
});
```

**Comment by sntmt** [comment]
Hello, I have the same problem, with the Go binding:

Here, I expected that webui.js is at http://localhost:5173/webui.js

```go
package main

import (
	"fmt"
	ui "github.com/webui-dev/go-webui/v2"
)

func greet(e ui.Event) string {
	name, _ := ui.GetArg[string](e)
	fmt.Printf("%s has reached the backend!\n", name)
	jsResp := fmt.Sprintf("Hello %s 🐇", name)
	return jsResp
}

func main() {
	w := ui.NewWindow()
	ui.Bind(w, "greet", greet)
	w.ShowBrowser("http://localhost:5173", ui.Firefox)
	ui.Wait()
}
```

Here is the index.html, which is served in the 5173 via vite:

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <titleLit component</title>
  <script src="/webui.js"></script>
  <script type="module" src="./my-element.js"></script>
</head>

<body>
  <my-element></my-element>
</body>

</html>
```

I can view the html inside the webui window, but the http://localhost:5173/webui.js do not exist.

**Reply by AlbertShown** [reply]
I assume you are using your own web server `http://localhost:5173`, then, in this case, please follow those steps:

- First, make sure your custom web server is running on `localhost:5173`
- In Go code, create a new WebUI window, and set a custom port like `5100`
- In HTML file, set `<script src="http://localhost:5100/webui.js"></script>`
- Show window `http://localhost:5173`

**Full working example in C:**
https://github.com/webui-dev/webui/tree/main/examples/C/custom_web_server

**Go Example:**

```html
<html>
  <head>
    <script src="http://localhost:5100/webui.js"></script>
  </head>
...
</html>
```

```go
func main() {
	w := ui.NewWindow()
	ui.SetPort(5100) // This mean `webui.js` will be at `localhost:5100/webui.js`
	w.ShowBrowser("http://localhost:5173", ui.Firefox)
	ui.Wait()
}
```
